Unit extensions API v2 for instructors dashboard

holaontiveros opened the pull request:
## Description

- Adds a new endpoint to get the unit_extensions for a course that will be used on the new instructor dashboard
- Modifies an utility function that needed to ignore values to unpack
- Updates edx-when from 3.0.0 to 3.1.0 that has the data required for the new endpoint to import


## Supporting information
fixes #37664 

## Testing instructions

- Make sure you have a valid course blocks that can be considered for an extension
- Use the example curl request to test

example curl to test, make sure you have the right cookies and update the right IDs and details
```bash
curl --location --request GET 'http://local.openedx.io:8000/api/instructor/v2/courses/course-v1:OEXCOM+101+2025_T1/unit_extensions' \
--header 'Accept: application/json, text/javascript, */*; q=0.01' \
--header 'Referer: http://local.openedx.io:8000/courses/course-v1:OEXCOM+101+2025_T1/instructor' \
--header 'Content-Type: application/json' \
--data-raw '{
    "block_id": "block-v1:OEXCOM+101+2025_T1+type@sequential+block@c3ccd3d1f91b42dfb2bff8185dba4d3b",
    "email_or_username": "javier",
    "due_datetime": "12/12/2025 12:12",
    "reason": "Just because"
}'
```

a possible valid example of response of the query looks like: 
```bash
{
    "next": null,
    "previous": null,
    "count": 2,
    "num_pages": 1,
    "current_page": 1,
    "start": 0,
    "results": [
        {
            "username": "armando",
            "full_name": "",
            "email": "armando.ontiveros@wgu.edu",
            "unit_title": "Subsection",
            "unit_location": "block-v1:OEXCOM+101+2025_T1+type@sequential+block@c3ccd3d1f91b42dfb2bff8185dba4d3b",
            "extended_due_date": "2026-12-28T00:00:00Z"
        },
        {
            "username": "javier",
            "full_name": "",
            "email": "javier.ontiveros@wgu.edu",
            "unit_title": "Subsection",
            "unit_location": "block-v1:OEXCOM+101+2025_T1+type@sequential+block@c3ccd3d1f91b42dfb2bff8185dba4d3b",
            "extended_due_date": "2025-12-12T12:12:00Z"
        }
    ]
}
```

## Deadline

None 

## Other information

Include anything else that will help reviewers and consumers understand the change.

- Requires https://github.com/openedx/edx-when/pull/345 to be released

holaontiveros committed:
chore: base work for unit extensions v2 api

holaontiveros committed:
chore: fix pylint things

holaontiveros committed:
chore: update to match edx-when comments

holaontiveros committed:
chore: update edx_when version

holaontiveros committed:
chore: restore old version before upgrade with make

holaontiveros committed:
chore: update with make upgrade-package package=edx-when

holaontiveros committed:
chore: exclude extras from get_overrides_for_block to prevent tuple from breaking

holaontiveros committed:
Merge branch 'master' into feature/unit_extensions_v2

wgu-taylor-payne commented on lms/djangoapps/instructor/views/tools.py:
```diff
@@ -220,7 +220,7 @@ def dump_block_extensions(course, unit):
     header = [_("Username"), _("Full Name"), _("Extended Due Date")]
     data = []
-    for username, fullname, due_date in api.get_overrides_for_block(course.id, unit.location):
+    for username, fullname, due_date, *unused in api.get_overrides_for_block(course.id, unit.location):
```
```suggestion
    for username, fullname, due_date, *_ in api.get_overrides_for_block(course.id, unit.location):
```

holaontiveros replied:
can't use _ because it's already being used for the translation things `_('thing_for_translation'` that's why I changed to something that was verbally undertandable

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -890,3 +891,410 @@ def test_get_graded_subsections_response_format(self):
+        # Create course structure with due dates
+        cls.chapter = BlockFactory.create(
+            parent=cls.course,
+            category='chapter',
+            display_name='Test Chapter'
+        )
+        cls.due_date = datetime(2024, 12, 31, 23, 59, 59, tzinfo=UTC)
+        cls.subsection = BlockFactory.create(
+            parent=cls.chapter,
+            category='sequential',
+            display_name='Homework 1',
+            due=cls.due_date
+        )
+        cls.vertical = BlockFactory.create(
+            parent=cls.subsection,
+            category='vertical',
+            display_name='Test Vertical'
+        )
+        cls.problem = BlockFactory.create(
+            parent=cls.vertical,
+            category='problem',
+            display_name='Test Problem'
+        )
```
Unless I missed something, other than the course/course_key, these other objects don't contribute to the tests. It looks like the two tests (`test_combined_filters` and `test_filter_by_block_id`) that return a non-empty results list are using mocks to supply the data. (I think you intended for more tests to return data, so I would suggest digging into these further to make sure we're testing what we're expecting).

I think that tests that use data setup in the database would be more valuable than mocking, so I would encourage switching over to that unless we can't avoid it.

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -890,3 +891,410 @@ def test_get_graded_subsections_response_format(self):
+        self.assertGreaterEqual(len(results), 0)
+
+        # If there are results, verify the structure
+        if results:
```
Do we expect data for this test? When I ran it `results` was empty, so I think we need more clarity on this test.

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -890,3 +891,410 @@ def test_get_graded_subsections_response_format(self):
+    @patch('lms.djangoapps.instructor.views.api_v2.get_units_with_due_date')
+    def test_filter_by_email_or_username(self, mock_get_units, mock_get_overrides):
+        """
+        Test filtering unit extensions by email or username.
```
In practice, this doesn't reach the part of the code that filters based on username/email, because it hits the conditional

```python
if location not in units_dict:
    continue
```

the conditional evaluates to `True` and it continues.

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -890,3 +891,410 @@ def test_get_graded_subsections_response_format(self):
+        # Test page parameter
+        params = {'page': '1', 'page_size': '10'}
+        url = f"{self._get_url()}?{urlencode(params)}"
+        response = self.client.get(url)
```
I believe you can just pass `params` as the second argument to `client.get`.

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -890,3 +891,410 @@ def test_get_graded_subsections_response_format(self):
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        data = response.data
+
+        if data['results']:
```
This didn't evaluate to `True` when I ran it. I think this test expects data, right? So we assert that results is not empty.

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -890,3 +891,410 @@ def test_get_graded_subsections_response_format(self):
+            self.assertIsInstance(extension['unit_title'], str)
+            self.assertIsInstance(extension['unit_location'], str)
+
+    def test_response_content_type(self):
```
I think we could do without this test.

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -890,3 +891,410 @@ def test_get_graded_subsections_response_format(self):
+        ('STUDENT1',),  # Test case insensitive
+        ('JOHN@EXAMPLE.COM',),  # Test case insensitive
+    )
+    @ddt.unpack
```
Since the tuples only have one value, we pass those values directly to `data` and then wouldn't need `ddt.unpack`.

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -890,3 +891,410 @@ def test_get_graded_subsections_response_format(self):
+        mock_get_overrides_block.assert_called_once()
+        mock_find_unit.assert_called_once()
+
+    def test_pagination_parameters(self):
```
It might be nice to test a scenario that returns more than one page of results.

wgu-taylor-payne commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -349,3 +352,179 @@ def get(self, request, course_id):
+                if email_or_username_filter:
+                    email_or_username_filter_lower = email_or_username_filter.lower()
+                    if (email_or_username_filter_lower not in username.lower() and
+                            email_or_username_filter_lower != email.lower()):
```
So, just to make sure the frontend is expecting partial matches on username and full match on email?

Also, why do these conditionals differ from the block-specific logic?

wgu-taylor-payne commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -349,3 +352,179 @@ def get(self, request, course_id):
+                return []
+        else:
+            # Use the course-wide logic from get_overrides_for_course
+            query_data = edx_when_api.get_overrides_for_course(course.id)
```
If `get_overrides_for_course` had the same return structure as `get_overrides_for_block`, I feel like we could clean up the duplication in the if/else blocks. Is it worth adding something (maybe a `get_overrides_for_course_v2`) to edx-when that does that? Since the two blocks are fairly close, are we able to combine them somehow?

dwong2708 replied:
I agree with the suggestion. 

We might need an adapter function like this o normalize both responses into the same structure:

```
def normalize_override_tuple(source, row):
    if source == "block":
        username, fullname, due_date, email, location = row
    elif source == "course":
        username, fullname, email, location, due_date = row
    else:
        raise ValueError("Unknown source")

    return {
        "username": username,
        "full_name": fullname,
        "email": email,
        "location": location,
        "due_date": due_date,
    }
```

and the usage could be:

```
query_data = edx_when_api.get_overrides_for_block(course.id, unit.location)
extension_data = [normalize_override_tuple("block", row) for row in query_data]
```

```
query_data = edx_when_api.get_overrides_for_course(course.id)
extension_data = [normalize_override_tuple("course", row) for row in query_data]
```

wgu-taylor-payne commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -349,3 +352,179 @@ def get(self, request, course_id):
+                        'extended_due_date': due_date,
+                    })
+
+            except (InvalidKeyError):
```
If `find_unit` is the only place this error is thrown, I would recommend just wrapping that line. Also, you can just write `except InvalidKeyError`, since there is only one exception we're catching here.

wgu-taylor-payne commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -349,3 +352,179 @@ def get(self, request, course_id):
+
+        return extension_data
+
+    def list(self, request, *args, **kwargs):
```
Perhaps rather than override [list](https://github.com/encode/django-rest-framework/blob/3f190b7ddc1ae19c9c9a3ed10f987c30b2eb9d5d/rest_framework/mixins.py#L38), we could just override `filter_queryset` to return the queryset passed into it.

wgu-taylor-payne replied:
Removing `list` and not overriding `filter_queryset` works without error because there are no [default filters](https://www.django-rest-framework.org/api-guide/settings/#default_filter_backends) in place. I think it would be good to explicitly set `filter_backends = []` on the class.

wgu-taylor-payne replied:
@dwong2708

dwong2708 replied:
Sorry, I didn't see this comment for some reason. I added filter_backends = []. Thanks

wgu-taylor-payne commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -349,3 +352,179 @@ def get(self, request, course_id):
+    permission_name = permissions.VIEW_DASHBOARD
+    serializer_class = UnitExtensionSerializer
+
+    @apidocs.schema(
```
It sounds like `apidocs.schema_for` ([docs](https://edx-api-doc-tools.readthedocs.io/en/stable/api.html#edx_api_doc_tools.schema_for)) might be more appropriate in this case.

wgu-taylor-payne commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -349,3 +352,179 @@ def get(self, request, course_id):
+    """
+    **Use Cases**
+
+        Retrieve a paginated list of due date extensions for units in a course.
```
It looks like the autogenerated swagger docs pick up the first line of the docstring by default, so this should be the first line rather than `**Use Cases**`.

dwong2708 commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -349,3 +352,179 @@ def get(self, request, course_id):
+                extension_data = []
+                for username, fullname, due_date, email, location in query_data:
+                    # Apply email_or_username filter if specified
+                    if email_or_username_filter and email_or_username_filter.lower() not in username.lower():
```
Was the email filter missed here?

dwong2708 committed:
Merge branch 'master' into feature/unit_extensions_v2

dwong2708 committed:
feat: address review comments

- Adjust tests for more consistent behavior
- Add UnitDueDateExtension as a data interface
- Reduce code duplication

dwong2708 committed:
Merge remote-tracking branch 'upstream/master' into feature/unit_extensions_v2

holaontiveros committed:
chore: fixed lint issue

dwong2708 committed:
Merge remote-tracking branch 'upstream/master' into feature/unit_extensions_v2

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -1126,3 +1129,478 @@ def test_permission_allowed_for_instructor(self):
+        data = response.data
+        results = data['results']
+
+        for r in results:
```
We only expect one result for each run, correct? In that case, looping wouldn't be necessary.

dwong2708 replied:
good catch. Changed. Thanks

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -1126,3 +1129,478 @@ def test_permission_allowed_for_instructor(self):
+            course_id = str(self.course_key)
+        return reverse('instructor_api_v2:unit_extensions', kwargs={'course_id': course_id})
+
+    def test_get_unit_extensions_success(self):
```
I think we could remove this one as the same things are covered in other tests.

dwong2708 replied:
done

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -1126,3 +1129,478 @@ def test_permission_allowed_for_instructor(self):
+
+    @patch('lms.djangoapps.instructor.views.api_v2.edx_when_api.get_overrides_for_course')
+    @patch('lms.djangoapps.instructor.views.api_v2.get_units_with_due_date')
+    def test_get_unit_extensions_with_data(self, mock_get_units, mock_get_overrides):
```
I think `test_get_unit_extensions_with_no_mockups` covers what we are tyring to test here, right?

dwong2708 replied:
right, I deleted it. Thanks

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -1126,3 +1129,478 @@ def test_permission_allowed_for_instructor(self):
+    def test_get_unit_extensions_with_no_mockups(self):
+        """
+        Test unit extensions with mocked data.
+        """
```
The docstring and method name seem to clash.

dwong2708 replied:
I changed to **test_get_unit_extensions**

wgu-taylor-payne commented on lms/djangoapps/instructor/tests/test_api_v2.py:
```diff
@@ -1126,3 +1129,478 @@ def test_permission_allowed_for_instructor(self):
+        self.assertEqual(len(results), 2)
+
+        # If there are results, verify the structure
+        extension = results[1]
```
I'm getting instructor data for `results[0]`. Is that expected?

dwong2708 replied:
results[1] is the expected

wgu-taylor-payne replied:
I'm trying to figure out exactly what we expect in this test result. We set up the test class:
```
course (course)
chapter (section)
sequential (subsection) - Due Dec 31, 2024  # I don't think this original due date is used
vertical (unit)
problem
```

Before each test we setup:
```
# set users
add instructor, staff to course
enroll student1 and student2

# set due dates for course elements
sequential - Mar 22, 2019
vertical - Mar 23, 2019
problem - Mar 24, 2019

# set extensions
instructor - subsection - April 1, 2019   # are extensions typically set for instructors or is this just contrived?
student1 - vertical - April 2, 2019
student1 - subsection - April 2, 2019
student2 - problem - April 3, 2019
```

So, we set up _four_ extensions on the course, and then we only get _two_ back in the response:

```python
{
  "next": null,
  "previous": null,
  "count": 2,
  "num_pages": 1,
  "current_page": 1,
  "start": 0,
  "results": [
    {
      "username": "robot0",
      "full_name": "Robot0 Instructor",
      "email": "robot+test+0@edx.org",
      "unit_title": "Homework 1",
      "unit_location": "block-v1:edX+TestX+Test_Course+type@sequential+block@Homework_1",
      "extended_due_date": "2019-04-01T00:00:00Z"
    },
    {
      "username": "student1",
      "full_name": "Robot2 Test",
      "email": "student1@example.com",
      "unit_title": "Homework 1",
      "unit_location": "block-v1:edX+TestX+Test_Course+type@sequential+block@Homework_1",
      "extended_due_date": "2019-04-02T00:00:00Z"
    }
  ]
}
```

Is that what we want? Coming into this test without much context around dates and extensions, it's a bit confusing to me.

dwong2708 replied:
I adjusted this important test to improve clarity. I also added comments to better explain the code. As far as I understand, the test should work as described below.

## Setup

### Course structure
- course
  - chapter (section)
    - vertical (unit)
      - problem

### Users
- Add instructor and staff to the course
- Enroll `student1` and `student2`

### Due dates
- sequential: Dec 31, 2024  
- vertical: Dec 31, 2024  
- problem: Dec 31, 2024  

### Extensions
- `student1` – subsection – **Oct 31, 2025**
- `student2` – subsection – Nov 30, 2025
- `student2` – subsection – **Dec 31, 2025**

## Expected behavior

The **first** and **third** extensions should be included in the resulting list, because only the **top-level unit with a due date** (the subsection) is considered when determining which extensions apply.

dwong2708 committed:
fix: apply review feedback to unit extensions tests

dwong2708 committed:
Merge remote-tracking branch 'upstream/master' into feature/unit_extensions_v2

wgu-taylor-payne commented on lms/djangoapps/instructor/views/api_v2.py:
```diff
@@ -356,6 +361,159 @@ def get(self, request, course_id):
+                    and filter_value not in unit_due_date_extension.email.lower()
+                ):
+                    continue
+            results.append(unit_due_date_extension)
```
Suggestion: extract the check to a function and then filter with a list comprehension.
```python
results = [
    extension
    for extension in unit_due_date_extensions
    if matches_email_or_username(ext, email_or_username_filter)
]
```

wgu-taylor-payne replied:
Another suggestion would be to only filter if we have a filter value:

```python
if filter_value:
    unit_due_date_extensions = # list comprehension
```

dwong2708 replied:
Done. Thanks.

wgu-taylor-payne reviewed (commented):
Here is one more suggestion I have.

dwong2708 committed:
refactor: simplify result filtering

dwong2708 committed:
Merge remote-tracking branch 'upstream/master' into feature/unit_extensions_v2

kdmccormick committed:
build: Update README, CODEOWNERS, catalog-info with new repo name (#37939)

Referenced GH team names have been updated accordingly as well.

Closes: https://github.com/openedx/openedx-platform/issues/37904

wgu-taylor-payne reviewed (commented):
Thanks for the update, I think that brings more clarity to the code. Could you please address these remaining items of feedback https://github.com/openedx/openedx-platform/pull/37783/#discussion_r2722874383 and https://github.com/openedx/openedx-platform/pull/37783/#discussion_r2710422094.

dwong2708 committed:
test: update test_get_unit_extensions

dwong2708 committed:
Merge branch 'master' into feature/unit_extensions_v2

wgu-taylor-payne reviewed (approved):
This looks good to me. Thanks @holaontiveros for the initial implementation and @dwong2708 for the feedback and improvements.

@feanil Were you wanting to give this one a look over?

feanil reviewed (approved):
Looks good to me as well.
